File: golang/battery.go
package main

type Battery struct {
	R0 float64
	R1 float64
	R2 float64
	C1 float64
	C2 float64
	I1 float64
	I2 float64
	I  float64
	Dt float64 // Time step
	Ni float64 // Coulombic Efficiency, 1 at discharge, Ni at charge
	Cn float64 // Nominal Capacity
	Zk float64 // State of Charge
}

// Initializes a new Battery.
func NewBattery(R0, R1, R2, C1, C2, Dt, Ni, Cn, Zk float64) *Battery {
	return &Battery{
		R0: R0,
		R1: R1,
		R2: R2,
		C1: C1,
		C2: C2,
		Dt: Dt,
		Ni: Ni,
		Cn: Cn,
		Zk: Zk,
	}
}

File: c/battery.c
#include <stdlib.h>

typedef struct {
	double R0;
	double R1;
	double R2;
	double C1;
	double C2;
	double I1;
	double I2;
	double I;
	double Dt; 
	double Ni; 
	double Cn;
	double Zk;
} Battery;

Battery* NewBattery(
	double R0,
	double R1,
	double R2,
	double C1,
	double C2,
	double I1,
	double I2,
	double I,
	double Dt, 
	double Ni, 
	double Cn,
	double Zk
) {
	Battery *b = NULL;

	b = malloc(sizeof(Battery));


	b->R0 = R0;
	b->R1 = R1;
	b->R2 = R2;
	b->C1 = C1;
	b->C2 = C2;
	b->I1 = I1;
	b->I2 = I2;
	b->I = I;
	b->Dt = Dt;
	b->Ni = Ni;
	b->Cn = Cn;
	b->Zk = Zk;

	return b;
};
